File: StudentGrade/StudentGrade.java
import java.util.Arrays;
import java.util.Scanner;
public class StudentGrade{
     static Scanner input =new Scanner(System.in);
     public static void main(String []args){

     System.out.print("How many students do you have ?: ");
     int noOfStudents = input.nextInt();
          
     System.out.print("How may subjects do they offer ?: ");
     int noOfSubjects = input.nextInt();
     System.out.println("Saving >>>>>>>>>>>>>>");
     System.out.println("Saved Successfully");

     int [] [] studentsResults = new int [noOfStudents][noOfSubjects];

          //   System.out.println(Arrays.deepToString(inputResults(studentsResults)));
          inputResults(studentsResults);
               
          int [] totalArray=  (totalPerStudent(studentsResults));
          double [] averageArray =  averagePerStudent(totalArray,noOfSubjects);
          int [] positionArray =    position(totalArray);
          int [] subjectsTotal=  totalPerSubject(studentsResults,noOfSubjects);
          double [] subjectAverageArray = subjectAverage(subjectsTotal,noOfStudents);
//          System.out.println(Arrays.toString(position(totalArray)));
          //     position(totalArray);
           System.out.println("=".repeat(70));

          System.out.print("STUDENT\t\t");
           for(int count=1;count<=noOfSubjects;count++){
               System.out.printf("SUB%d \t",count);
           }
          System.out.println("TOT\t Ave\t POS" );
           for (int index=0;index <noOfStudents;index++){
               System.out.printf("Student %d\t",(index+1));
              for (int element =0;element <noOfSubjects;element++){
             System.out.print(studentsResults[index][element]);
             System.out.print("\t");
          //    System.out.print(totalArray
              }
               System.out.print(totalArray[index]);
             System.out.print("\t");
              System.out.printf("%.2f",averageArray[index]);
             System.out.print("\t");
              System.out.printf("%d",positionArray[index]);
              System.out.println();
           }
           System.out.println("=".repeat(70));
           System.out.println();


           System.out.println("SUBJECT SUMMARY");
          //  for(int position=0 ;position<noOfSubjects;position++){
                         subjectSummary(studentsResults, noOfStudents, noOfSubjects,subjectsTotal,subjectAverageArray);
                         hardestAndEasiest (studentsResults);
                         classSummary( totalArray,noOfStudents);
          //                System.out.printf("Total Score is %d%n",subjectsTotal[position]);
          //                System.out.printf("Average Score is: %.2f",subjectAverageArray[position]);
          // //  }

               // totalPerSubject(studentsResults,noOfSubjects);
               // subjectAverage((totalPerSubject(studentsResults,noOfSubjects)),noOfStudents);
}
//          {{1,2,1},{2,3,4},{2,5,7}}
     public static int [][]  inputResults(int [][] studentsResults){


     for(int count=0; count<studentsResults.length;count++){

          for (int index=0; index<studentsResults[count].length; index++){
                    System.out.println("Entering Score for Student  "+(count+1));
                    System.out.println("Enter Score for Subject  "+(index+1));
                    studentsResults[count][index] = input.nextInt();
                    System.out.println("Saving >>>>>>>>>>>>>>");
                    System.out.println("Saved Successfully");
                    System.out.println();          
}
}
             return studentsResults;
   }

     public static  int [] totalPerStudent (int [][] studentsResults){
     
     int [] totalArray = new int [studentsResults.length]; 

     for(int count=0; count<totalArray.length;count++){

          for (int index=0; index<studentsResults[count].length; index++){

          totalArray[count]+= studentsResults[count][index];
}
}

          return totalArray;
}

 
     public static double [] averagePerStudent (int [] totalArray,int noOfSubjects){
     double [] averageArray = new double   [totalArray.length];
     
     for(int count=0; count<averageArray.length;count++){
          
       averageArray[count] =((double) totalArray[count]) /noOfSubjects;
     
}

     return averageArray;

}

      public static int [] position (int  [] totalArray){

     int [] positionArray = new int [totalArray.length];

     for (int count =0; count <totalArray.length;count++){
          int position =1;
          for(int index=0; index<totalArray.length;index++){
          
          if(totalArray[index]>totalArray[count]){
               position++;    
}

}     

               positionArray[count]=position;
}
     // System.out.println(Arrays.toString(positionArray));
               return positionArray;
} 


public static int [] totalPerSubject (int [][] studentsResults, int noOfSubjects){
               int [] subjectsTotal = new int[noOfSubjects];

     for(int count=0;count <studentsResults.length;count++){

          for(int index=0;index<subjectsTotal.length;index++){
               subjectsTotal [index] += studentsResults[count] [index];
               
          }

     }
          // System.out.println(Arrays.toString(subjectsTotal));

          return  subjectsTotal;
}

public static double [] subjectAverage (int [] subjectsTotal,int noOfStudents){

     double [] subjectAverageArray =new double [subjectsTotal.length];

          for (int count =0;count<subjectAverageArray.length;count++){
               subjectAverageArray[count]=((double)subjectsTotal[count])/ noOfStudents;
          }
               // System.out.println(Arrays.toString(subjectAverageArray));
               return subjectAverageArray;
}

public static void subjectSummary(int [] [] studentsResults,int noOfStudents, int noOfSubjects, int [] subjectsTotal, double[] subjectAverageArray) {
     
     for(int count =0;count<noOfSubjects;count++){
          int highestScore = studentsResults[0][count];
          int lowestScore = studentsResults[0][count];
          int highestScoringStudent = 1;
          int lowestScoringStudent = 1;
          int pass =0;
          int fail =0;

               for (int index=0;index<noOfStudents;index++){
                    int result = studentsResults[index][count];

                    if(result >highestScore){
                         highestScore=result;
                         highestScoringStudent = index+1;
                    }
                    if (result <lowestScore){
                         lowestScore = result;
                         lowestScoringStudent = index+1;
                    }
                     if (result>= 50){
                         pass++;
                     }
                     else{
                         fail++;
                     }

               }


                     System.out.println();
                     System.out.println("Subject "+(count+1));
                     System.out.printf("Highest scoring Student is: Student %d scoring %d%n",highestScoringStudent,highestScore );
                     System.out.printf("Lowest scoring Student is: Student %d scoring %d%n",(lowestScoringStudent),lowestScore );
                     System.out.println("Number of passes: "+pass);
                     System.out.println("Number of fails: "+fail);
                       System.out.printf("Total Score is %d%n",subjectsTotal[count]);
                        System.out.printf("Average Score is: %.2f",subjectAverageArray[count]);
                        System.out.println();

     }
}

public static void classSummary (int[] totalArray,int noOfStudents){
     int sum =0;
     int highest =totalArray[0];
     int lowest = totalArray[0];
     int highestScoringStudent = 1 ;
     int lowestScoringStudent = 1;
     for (int count=0;count<noOfStudents;count++){
           int result = totalArray[count];
               sum += result;
                    if(result >highest){
                         highest=result;
                         highestScoringStudent = count+1;
                    }
                    if (result <lowest){
                         lowest = result;
                         lowestScoringStudent = count+1;
                    }


     }

     double classAverage =((double)sum)/noOfStudents;
     System.out.println();
     System.out.println("CLASS SUMMARY");
     System.out.println("=".repeat(50));
     System.out.printf("Best Graduating Student is: Student %d scoring %d%n",highestScoringStudent,highest);
     System.out.println("=".repeat(50));
     System.out.println();
     System.out.println("!".repeat(50));
     System.out.printf("Worst Graduating Student is: Student %d scoring %d%n",(lowestScoringStudent),lowest);
     System.out.println("!".repeat(50));
     System.out.println();
     System.out.println("=".repeat(50));
     System.out.println("Class Total Score is: "+sum);
     System.out.printf("Class Average is: %.2f%n", classAverage);




}

public static void hardestAndEasiest (int [][] studentsResults){
     int highestScore = studentsResults[0][0];
     int lowestScore = studentsResults[0][0];
     int highestScoringStudent =1;
     int lowestScoringStudent =1;
     int highestSubject=1;
     int lowestSubject=1;
     int [] passArray = new int [studentsResults[0].length];
     // int highPass = passArray[0];
     // int highPassSubject = 0;
     // int lowPass = passArray[0];
     // int lowPassSubject=0; 

     for (int count =0;count< studentsResults[0].length;count++){
          
          for(int index=0;index<studentsResults.length;index++){
               int result =studentsResults[index][count];

                if (result>= 50){
                         passArray[count] +=1;
                     }
               if(result >highestScore){
                         highestScore=result;
                         highestScoringStudent = index+1;
                         highestSubject=count+1;

                    }
                    if(result <lowestScore){
                         lowestScore = result;
                         lowestScoringStudent = index+1;
                         lowestSubject= count+1;
                    }
          }
     }

     int highPass = passArray[0];
     int highPassSubject = 1;
     int lowPass = passArray[0];
     int lowPassSubject=1; 


          for (int element =0;element<passArray.length;element++){
               int pass = passArray[element];
               if(pass >highPass){
                         highPass=pass;
                         highPassSubject = element+1;

                    }
                    if (pass <lowPass){
                         lowPass = pass;
                         lowPassSubject = element+1;
                         
          }
          }
                    int failures = studentsResults.length - lowPass;

          System.out.println();
          System.out.printf("The Hardest subject is subject %d with %d failures%n",lowPassSubject,failures);
          System.out.printf("The Easiest subject is subject %d with %d passes%n",highPassSubject,highPass);

          System.out.printf("The overall Highest Score is scored by Student %d in subject %d scoring %d%n",highestScoringStudent,highestSubject,highestScore);
          System.out.printf("The overall lowest Score is scored by Student %d in subject %d scoring %d%n",lowestScoringStudent,lowestSubject,lowestScore);
          System.out.println();
}

}

File: StudentGrade/Student_Grade.py

def input_results(students_results):
    for student_index in range(len(students_results)):
        for subject_index in range(len(students_results[student_index])):
            students_results[student_index][subject_index] = int(
                input(
                    f"Entering Score for Student {student_index + 1}\n"
                    f"Enter Score for Subject {subject_index + 1}: "
                )
            )
            print("Saving >>>>>>>>>>>>>>")
            print("Saved Successfully")
            print()
    return students_results


def total_per_student(students_results):
    total_array = [0] * len(students_results)

    for student_index in range(len(students_results)):
        for subject_index in range(len(students_results[student_index])):
            total_array[student_index] += students_results[student_index][subject_index]

    return total_array


def average_per_student(total_array, no_of_subjects):
    average_array = [0.0] * len(total_array)

    for index in range(len(total_array)):
        average_array[index] = total_array[index] / no_of_subjects

    return average_array


def position(total_array):
    position_array = [0] * len(total_array)

    for current in range(len(total_array)):
        pos = 1
        for other in range(len(total_array)):
            if total_array[other] > total_array[current]:
                pos += 1
        position_array[current] = pos

    return position_array


def total_per_subject(students_results, no_of_subjects):
    subjects_total = [0] * no_of_subjects

    for student_index in range(len(students_results)):
        for subject_index in range(no_of_subjects):
            subjects_total[subject_index] += students_results[student_index][subject_index]

    return subjects_total


def subject_average(subjects_total, no_of_students):
    subject_average_array = [0.0] * len(subjects_total)

    for index in range(len(subjects_total)):
        subject_average_array[index] = subjects_total[index] / no_of_students

    return subject_average_array


def subject_summary(
    students_results,
    no_of_students,
    no_of_subjects,
    subjects_total,
    subject_average_array
):
    for subject_index in range(no_of_subjects):
        highest_score = students_results[0][subject_index]
        lowest_score = students_results[0][subject_index]
        highest_scoring_student = 1
        lowest_scoring_student = 1
        passes = 0
        fails = 0

        for student_index in range(no_of_students):
            result = students_results[student_index][subject_index]

            if result > highest_score:
                highest_score = result
                highest_scoring_student = student_index + 1

            if result < lowest_score:
                lowest_score = result
                lowest_scoring_student = student_index + 1

            if result >= 50:
                passes += 1
            else:
                fails += 1

        print()
        print(f"Subject {subject_index + 1}")
        print(f"Highest scoring Student is: Student {highest_scoring_student} scoring {highest_score}")
        print(f"Lowest scoring Student is: Student {lowest_scoring_student} scoring {lowest_score}")
        print("Number of passes:", passes)
        print("Number of fails:", fails)
        print(f"Total Score is {subjects_total[subject_index]}")
        print(f"Average Score is: {subject_average_array[subject_index]:.2f}")
        print()


def class_summary(total_array, no_of_students):
    total_sum = 0
    highest_score = total_array[0]
    lowest_score = total_array[0]
    highest_scoring_student = 1
    lowest_scoring_student = 1

    for index in range(no_of_students):
        result = total_array[index]
        total_sum += result

        if result > highest_score:
            highest_score = result
            highest_scoring_student = index + 1

        if result < lowest_score:
            lowest_score = result
            lowest_scoring_student = index + 1

    class_average = total_sum / no_of_students

    print()
    print("CLASS SUMMARY")
    print("=" * 50)
    print(f"Best Graduating Student is: Student {highest_scoring_student} scoring {highest_score}")
    print("=" * 50)
    print()
    print("!" * 50)
    print(f"Worst Graduating Student is: Student {lowest_scoring_student} scoring {lowest_score}")
    print("!" * 50)
    print()
    print("=" * 50)
    print("Class Total Score is:", total_sum)
    print(f"Class Average is: {class_average:.2f}")


def hardest_and_easiest(students_results):
    highest_score = students_results[0][0]
    lowest_score = students_results[0][0]
    highest_scoring_student = 1
    lowest_scoring_student = 1
    highest_subject = 1
    lowest_subject = 1

    pass_array = [0] * len(students_results[0])

    for subject_index in range(len(students_results[0])):
        for student_index in range(len(students_results)):
            result = students_results[student_index][subject_index]

            if result >= 50:
                pass_array[subject_index] += 1

            if result > highest_score:
                highest_score = result
                highest_scoring_student = student_index + 1
                highest_subject = subject_index + 1

            if result < lowest_score:
                lowest_score = result
                lowest_scoring_student = student_index + 1
                lowest_subject = subject_index + 1

    highest_pass = pass_array[0]
    lowest_pass = pass_array[0]
    highest_pass_subject = 1
    lowest_pass_subject = 1

    for index in range(len(pass_array)):
        if pass_array[index] > highest_pass:
            highest_pass = pass_array[index]
            highest_pass_subject = index + 1

        if pass_array[index] < lowest_pass:
            lowest_pass = pass_array[index]
            lowest_pass_subject = index + 1

    failures = len(students_results) - lowest_pass

    print()
    print(f"The Hardest subject is subject {lowest_pass_subject} with {failures} failures")
    print(f"The Easiest subject is subject {highest_pass_subject} with {highest_pass} passes")
    print(f"The overall Highest Score is scored by Student {highest_scoring_student} "f"in subject {highest_subject} scoring {highest_score}")
    print(f"The overall lowest Score is scored by Student {lowest_scoring_student} "f"in subject {lowest_subject} scoring {lowest_score}")
    print()




no_of_students = int(input("How many students do you have ?: "))
no_of_subjects = int(input("How many subjects do they offer ?: "))

print("Saving >>>>>>>>>>>>>>")
print("Saved Successfully")


students_results = []
for _ in range(no_of_students):
    row = [0] * no_of_subjects
    students_results.append(row)


input_results(students_results)
total_array = total_per_student(students_results)
average_array = average_per_student(total_array, no_of_subjects)
position_array = position(total_array)
subjects_total = total_per_subject(students_results, no_of_subjects)
subject_average_array = subject_average(subjects_total, no_of_students)

print("=" * 70)

print("STUDENT\t\t", end="")
for count in range(1, no_of_subjects + 1):
    print(f"SUB{count}\t", end="")
print("TOT\t Ave\t POS")

for index in range(no_of_students):
    print(f"Student {index + 1}\t", end="")
    for element in range(no_of_subjects):
        print(students_results[index][element], end="\t")
    print(total_array[index], end="\t")
    print(f"{average_array[index]:.2f}", end="\t")
    print(position_array[index])

print("=" * 70)
print()

print("SUBJECT SUMMARY")
subject_summary(students_results,no_of_students,no_of_subjects,subjects_total,subject_average_array)
hardest_and_easiest(students_results)
class_summary(total_array, no_of_students)
